pkg/util/informermanager/informermanager.go

```diff
@@ -0,0 +1,287 @@
+
+	if oldGVR, exists := m.gvrMapping.Lookup(ftcName); exists {
+		if oldGVR != gvr {
+			// This might occur if a ftc was deleted and recreated with a different source type within a short period of
+			// time and we missed processing the deletion. We simply process the ftc deletion and reenqueue. Note:
+			// updating of ftc source types, however, is still not a supported use case.
+			err := m.processFTCDeletionUnlocked(ftcName)
+			return err, true
+		}
+
+		informer = m.informers[ftcName]
+	} else {
+		if err := m.gvrMapping.Add(ftcName, gvr); err != nil {
+			// There must be another ftc with the same source type GVR.
+			return fmt.Errorf("source type is already referenced by another FederatedTypeConfig: %w", err), false
+		}
+
+		informer = dynamicinformer.NewFilteredDynamicInformer(
+			m.client,
+			gvr,
+			metav1.NamespaceAll,
+			0,
+			cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+			nil,
+		)
+		stopCh := make(chan struct{})
+		go informer.Informer().Run(stopCh)
+
+		m.informers[ftcName] = informer
+		m.informerStopChs[ftcName] = stopCh
+		m.eventHandlerRegistrations[ftcName] = map[*EventHandlerGenerator]cache.ResourceEventHandlerRegistration{}
+		m.lastAppliedFTCsCache[ftcName] = map[*EventHandlerGenerator]*fedcorev1a1.FederatedTypeConfig{}
+	}
+
+	registrations := m.eventHandlerRegistrations[ftcName]
+	lastAppliedFTCs := m.lastAppliedFTCsCache[ftcName]
+
+	ftc = ftc.DeepCopy()
+
+	for _, generator := range m.eventHandlerGenerators {
+		lastApplied := lastAppliedFTCs[generator]
+		if !generator.Predicate(lastApplied, ftc) {
+			lastAppliedFTCs[generator] = ftc
+			continue
+		}
+
+		if oldRegistration := registrations[generator]; oldRegistration != nil {
+			if err := informer.Informer().RemoveEventHandler(oldRegistration); err != nil {
+				return fmt.Errorf("failed to unregister event handler: %w", err), true
+			}
+			delete(registrations, generator)
+		}
+		delete(lastAppliedFTCs, generator)
+
+		if handler := generator.Generator(ftc); handler != nil {
+			newRegistration, err := informer.Informer().AddEventHandler(handler)
+			if err != nil {
+				return fmt.Errorf("failed to register event handler: %w", err), true
+			}
+			registrations[generator] = newRegistration
+		}
+
+		lastAppliedFTCs[generator] = ftc
+	}
+
+	return nil, false
+}
+
+func (m *informerManager) processFTCDeletion(ftcName string) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	return m.processFTCDeletionUnlocked(ftcName)
+}
+
+func (m *informerManager) processFTCDeletionUnlocked(ftcName string) error {
+	if stopCh, ok := m.informerStopChs[ftcName]; ok {
+		close(stopCh)
+	}
+
+	m.gvrMapping.Delete(ftcName)
+
+	delete(m.informers, ftcName)
+	delete(m.informerStopChs, ftcName)
+	delete(m.eventHandlerRegistrations, ftcName)
+
+	return nil
+}
+
+func (m *informerManager) AddEventHandlerGenerator(generator *EventHandlerGenerator) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.started {
```

Comment:
Does this mean the informer manager can only be started after all the controllers have registered their event handler generators?

Reply:
yes, that is right.